### PR TITLE
Add sum of squares of weights to collapse output

### PR DESCRIPTION
**Pull request by mwilensky768**
## Description

This PR adds an output to `utils.collapse`, specifically `mean_collapse`, `absmean_collapse`, and `quadmean_collapse`, to return the sum of the squares of the weights on collapsing. 

## Motivation and Context

When the weights are not all equal to 1 or 0, this is necessary to standardize collapsed data and is not retrievable from what is returned already. SSINS, which subclasses UVFlag, will require this to make use of unevenly weighted data. I imagine other applications of UVFlag will as well.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change

## Checklist:

- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

Bug fix checklist:
- [ ] My fix includes a new test that breaks as a result of the bug (if possible).
- [ ] All new and existing tests pass in both python 2 and python 3.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass in both python 2 and python 3.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

Breaking change checklist:  
- [ ] I have updated the docstrings associated with my change using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to reflect my changes (if appropriate).
- [ ] My change includes backwards compatibility and deprecation warnings (if possible).
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests pass in both python 2 and python 3.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

Documentation change checklist:
- [ ] Any updated docstrings use the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] If this is a significant change to the readme or other docs, I have checked that they are rendered properly on ReadTheDocs. (you may need help to get this branch to build on RTD, just ask!)

Version change checklist:
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md) to put all the unreleased changes under the new version (leaving the unreleased section empty).
- [ ] I have noted any dependency changes since the last version and will update the conda package build accordingly.

Build or continuous integration change checklist:
- [ ] If required or optional dependencies have changed (including version numbers), I have updated the readme to reflect this.
- [ ] If this is a new CI setup, I have added the associated badge to the readme and to references/make_index.py (if appropriate).

### CODE REVIEW

**Comment by mwilensky768** [comment]
Currently I've just modified the functions in `utils` to return the item I want returned. I still need to change UVFlag's `waterfall` method to actually return this item. I'm wondering about whether it should be a default attribute of UVFlag objects (I think it has a range of uses), or if it should be optional. Thoughts appreciated.

**Comment by mkolopanis** [comment]
I just want to make sure I understand. You're interested in adding a `weights_squared` parameter to `UVFlag`? If it is a useful quantity to carry around then I am totally for it. While I am the primary maintainer of `UVFlag` now I honestly am not too familiar with the uses for things like this. It would be great to have someone more hands on involved.

**Comment by mwilensky768** [comment]
@mkolopanis yes something like that. I can try to post a couple lines of math justifying the need for such a quantity if desired. In English, a weighted sum of independent random variables (as implemented in `mean_collapse`, which I'm of the firm opinion is the correct way to do it) with identical variance will sum to a new random variable with variance proportional to the original variance, where the constant of proportionality is equal to the sum of the weights^2 divided by the square of the sum of the weights.

**Comment by adampbeardsley** [comment]
So I think you want the plumbing to do that operation, but you don't necessarily need both weights and weights_squared around, right? I don't want to argue use cases -- if you really want both that's fine -- but I would think the new output weights being 1/var(output data) would be what you want.

**Comment by adampbeardsley** [comment]
Also, for a little bit of context:

Returning the sum of the weights was chosen under the assumption that weights are typically 1/variance (for a minimum noise mean). Under that assumption, the resulting variance is 1/sum(weights), so if you weight by 1/variance again, the new weights are just the sum of the original weights.

Again, if you're doing something different, that's fine and we can figure out how to support it. Just want to be sure you're aware of what's there already.

**Comment by mwilensky768** [comment]
I think keeping the sum of the squares extends code that is already in place to handle the case of general weights, which is a nice addition. I will spare you my thoughts on higher order cumulants (skewness etc.) since you do not want to argue about use cases.

**Comment by adampbeardsley** [comment]
Ok, then I think my biggest concern is the naming. We previously used `return_weights` because, although it was the sum of the input weights, it was interpreted as the weights of the output.
But `return_weights_square` would seem to imply it's returning the square of the output weights.

Could we change them to be `return_weight_sum` and `return_weight_square_sum`? Is that too verbose for keywords? Would also require an API change. What does @mkolopanis think?

**Comment by mwilensky768** [comment]
Does `return_square_weights` do any better? I feel like it's more confusing. If so, I think your suggestions are fine. API changes I guess can be icky for backwards compatibility. Not sure what is worse here.

**Comment by adampbeardsley** [comment]
I think I was just overthinking this. Let's go with your original notation, and just make sure it's clear in docstrings (which I think it is).

Could you update the CHANGELOG? Once that's done, this should be good to go.

For UVFlag, I think it should be an optional parameter. I don't think most users will use it, and it would be confusing if you don't understand its uses.

**Comment by mkolopanis** [comment]
This solution will work for me I think.

**Comment by mwilensky768** [comment]
It seems like the process of adding an attribute to a UVFlag object, optional or not, is somewhat involved, but I think it is the right thing to do in this case. Is there a checklist of things to do when adding a parameter to a UVFlag object?

**Comment by mwilensky768** [comment]
Ideally this would only be added as an attribute when the `type` was converted to `waterfall`

**Comment by bhazelton** [comment]
This needs to be rebased on master to help tests pass.

**Comment by bhazelton** [comment]
We're going to try to get this in and then add it as an optional parameter on UVFlag in a subsequent PR.

File: pyuvdata/utils.py [hunk 1]
```diff
@@ -1496,16 +1512,24 @@ def quadmean_collapse(arr, weights=None, axis=None, return_weights=False):
...
-    if return_weights:
+    out = mean_collapse(np.abs(arr)**2, weights=weights, axis=axis,
+                        return_weights=return_weights,
+                        return_weights_square=return_weights_square)
+    if return_weights and return_weights_square:
+        return np.sqrt(out[0]), out[1], out[2]
+    elif return_weights or return_weights_square:
         return np.sqrt(out[0]), out[1]
+    elif return_weights_square:
+        return np.sqrt(out[0]), out[2]
```
Comment (bhazelton) [inline pyuvdata/utils.py]: It looks like this line isn't currently being covered by tests.

Comment (mwilensky768) [reply pyuvdata/utils.py]: Thanks! It looks like I managed to fix this.

**Pull request merged**